src/epoch.rs

```diff
@@ -27,49 +26,39 @@ impl Epoch {
         Self::default()
     }
 
-    /// Returns the number of steps between two epochs.
-    ///
-    /// An epoch is internally represented as an integer that wraps around at some unspecified
-    /// point. The returned distance between two epochs is the minimum number of steps required to
-    /// go from `self` to `other` or go from `other` to `self`.
-    #[inline]
-    pub fn distance(&self, other: Self) -> usize {
-        let (e1, _) = self.decompose();
-        let (e2, _) = other.decompose();
-        cmp::min(e1.wrapping_sub(e2), e2.wrapping_sub(e1))
+    /// Returns the number of epochs `self` is ahead of `rhs`.
+    /// 
+    /// Internally, epochs are represented as numbers in the range `(isize::MIN / 2) .. (isize::MAX
+    /// / 2)`, so the returned distance will be in the same interval.
+    pub fn wrapping_sub(self, rhs: Self) -> isize {
+        (self.data & !1).wrapping_sub(rhs.data & !1) as isize >> 1
```

Comment:
I realized you probably don't need the first (`& !1`) because it's shifted away. The second one has to stay.

Reply:
Right.  I expected LLVM InstCombine will optimize it out, but I could confirm that it didn't.

I pushed a patch that removes the first `& !1`.  But I'm worrying about the readability: even though I left a comment, it may be strange to someone's eyes.  Any thoughts?  cc @stjepang